### DISCUSSION TITLE
Javascript indexes and JS Date helper methods ?

### DISCUSSION THREAD

**Discussion by mcrio** [ravendb/ravendb#15363] [Q&A] [answered]
Hi, are there some helper methods to align the serialisation of the javascript Date object with the serialisation rules used when writing C# indexes and queries.

I am writing a javascript index as Jint is having problems transforming a particular LINQ statement and have an issues with JS Date.

Goal: Using JS, index (map/reduce) just the Date portion of `2022-11-24T05:00:01.0000000+01:00` as we are interested in the Date part only.

1. C#  `myDateTime.Date` indexes the value as `2022-11-24T05:00:01.0000000`
2. JS `myDate` indexes the value as `2022-11-24t00:00:00.0000000z`
3. JS `myDate.toISOString().toUpperCase().slice(0, -1)` indexes the value as `2022-11-24t00:00:00.000` (note the lowercase t)

Querying using the C# client produces a query like `...where MyDate = '2022-11-24T00:00:00.000'`  This query will not match anything due to the uppercase `T`.

In order for this to work I need to use the 3rd approach and explicitly set `MyDate` field indexing to `Exact` so `T` is not lowercased.

Any easier way to make this work?

**Comment by mcrio** [comment]
Actually the above mentioned toISOString will have 3 decimals precision while we need 7 with the indexing Exact method.
So we could add  additional sources to define a method to do the transformation:
```
AdditionalSources = new Dictionary<string, string>
        {
            ["Script"] = @"var __toDatePortion = date => {
  var pad = (n) => (Math.floor(Math.abs(n)) +'').padStart(2, '0');
  return date.getFullYear() +
    '-' + pad(date.getMonth() + 1) +
    '-' + pad(date.getDate()) +
    'T00:00:00.0000000'; 
};",
        };
```

.. and use this method when indexing the date only part of a JS Date object.

I really believe I'm doing something wrong and this could be done in a more optimal way :)

**Comment by garayx** [comment, marked as answer]
try to put `myDate` inside the `new Date()` constructor for example `OrderedAt: new Date(order.OrderedAt)`

**Reply by mcrio** [reply]
Hi @garayx Unfortunately as described above that will just use the default serialisation format (`2022-11-24t00:00:00.0000000z`) and include the timezone identifier as well, so won't match against a C# `myDateTime.Date` which is trying to match `2022-11-24T00:00:00.0000000`.

The only solution for me was that custom serializer method + setting the field indexing type to Exact. I guess there must be some better alternative.

**Reply by georgiosd** [reply]
Why not just use UTC ISO on both sides and cut-off the three extra zeros on the index?

**Reply by mcrio** [reply]
Hi. The document actually holds a DateTimeOffset value but I'm indexing and querying the Date part only. (Haven't upgraded the project yet to support DateOnly)

RavenDb by default will serialize C# DateTime to ISO8601 but with 7 digits precision.

I guess I could go ahead and set UTC for the `myDateTimeOffset.Date` c# object (when querying), but on index side I'd still have to set Index to`Exact` for that field to preserve the uppercase T as it wasn't matching when `T` vs `t`, and extend the precision to 7 digits.

I was just wondering if there are some RavenDb built in helper methods for JS to seamlessly handle the dates.

**Reply by garayx** [reply]
by using new Date() constructor I didn't get a `t` & `z` (lowercase), but got: `2022-11-24T00:00:00.0000000Z` (uppercase), can you share a [unit test](https://ravendb.net/docs/article-page/5.1/Csharp/start/test-driver) to demonstrate this?

**Reply by georgiosd** [reply]
C# will still parse the JS milisecond precision! This runs succesfully: `DateTime.Parse(DateTime.UtcNow.Date.ToString("o").Remove(24, 3))`

**Reply by mcrio** [reply]
> ..can you share a [unit test](https://ravendb.net/docs/article-page/5.1/Csharp/start/test-driver) to demonstrate this?

Please try this one: https://gist.github.com/mcrio/0b76f7938f3f8e3f4fe6afa69aedac82

**Reply by garayx** [reply]
Thanks, it seems when it is a map-reduce index the terms are converted to lowercase. In only map index the terms are fine.  You can remove the reduce in `Person_BirthDayIndex` index, and the term will be uppercase. I have opened an issue to fix this

https://issues.hibernatingrhinos.com/issue/RavenDB-19625/

**Reply by mcrio** [reply]
Thanks. Also to note that the suggested (marked) answer of using `new Date()` does not produce the expected result as it'll add the timezone identifier as well, while C# query does not. The test case shows it.
